tests/Unit/Product/ProductFactoryTest.php

```diff
@@ -0,0 +1,116 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\GoogleListingsAndAds\Tests\Unit\Product;
+
+use Automattic\WooCommerce\GoogleListingsAndAds\Exception\InvalidClass;
+use Automattic\WooCommerce\GoogleListingsAndAds\Product\Attributes\AttributeManager;
+use Automattic\WooCommerce\GoogleListingsAndAds\Product\ProductFactory;
+use Automattic\WooCommerce\GoogleListingsAndAds\Product\WCProductAdapter;
+use Automattic\WooCommerce\GoogleListingsAndAds\Proxies\WC;
+use Automattic\WooCommerce\GoogleListingsAndAds\Tests\Framework\ContainerAwareUnitTest;
+use Automattic\WooCommerce\GoogleListingsAndAds\Tests\Tools\HelperTrait\ProductTrait;
+use PHPUnit\Framework\MockObject\MockObject;
+use WC_Helper_Product;
+use WC_Product;
+
+/**
+ * Class ProductFactoryTest
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\Tests\Unit\Product
+ *
+ * @property MockObject|AttributeManager $attribute_manager
+ * @property WC                          $wc
+ * @property ProductFactory              $product_factory
+ */
+class ProductFactoryTest extends ContainerAwareUnitTest {
+
+	use ProductTrait;
+
+	public function test_create() {
+		$product = WC_Helper_Product::create_simple_product();
+
+		$attributes = $this->get_sample_attributes();
+		$this->attribute_manager->expects( $this->any() )
+								->method( 'get_all_values' )
+								->willReturn( $attributes );
+
+		$adapted_product = $this->product_factory->create( $product, 'US' );
+
+		$this->assertInstanceOf( WCProductAdapter::class, $adapted_product );
+
+		// make sure the target country is set
+		$this->assertEquals( 'US', $adapted_product->getTargetCountry() );
+
+		// check if attribute are all set in the created product
+		$this->assertEquals( $attributes['gtin'], $adapted_product->getGtin() );
+		$this->assertEquals( $attributes['mpn'], $adapted_product->getMpn() );
+		$this->assertEquals( $attributes['brand'], $adapted_product->getBrand() );
+		$this->assertEquals( $attributes['condition'], $adapted_product->getCondition() );
+		$this->assertEquals( $attributes['gender'], $adapted_product->getGender() );
+		$this->assertContains( $attributes['size'], $adapted_product->getSizes() );
+		$this->assertEquals( $attributes['sizeSystem'], $adapted_product->getSizeSystem() );
+		$this->assertEquals( $attributes['sizeType'], $adapted_product->getSizeType() );
+		$this->assertEquals( $attributes['color'], $adapted_product->getColor() );
+		$this->assertEquals( $attributes['material'], $adapted_product->getMaterial() );
+		$this->assertEquals( $attributes['pattern'], $adapted_product->getPattern() );
+		$this->assertEquals( $attributes['ageGroup'], $adapted_product->getAgeGroup() );
+		$this->assertEquals( $attributes['multipack'], $adapted_product->getMultipack() );
+		$this->assertEquals( $attributes['isBundle'], $adapted_product->getIsBundle() );
+		$this->assertEquals( $attributes['adult'], $adapted_product->getAdult() );
```

Comment:
This seems to work great with testing products with "valid" attributes. How about all the validation rules we setup in the function `load_validator_metadata`? Do we need to create any tests for invalid attributes?

Reply:
Yeah, we definitely need to test those but I thought it would be best to test it as part of the `WCProductAdapter` unit test. Here the main idea is to check whether the attributes are passed to the `WCProductAdapter` object that is created.